File: backend/app/models/product.py
"""
Product model — the central entity in the PIM system.

Each product belongs to a category and a segment, and tracks its position
in the content marketing workflow via a Kanban-style stage enum.
"""

import enum
from datetime import datetime

from sqlalchemy import (
    Boolean,
    DateTime,
    Enum,
    ForeignKey,
    Integer,
    String,
    Text,
    func,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class WorkflowStage(str, enum.Enum):
    """Kanban workflow stages for content marketing pipeline."""

    INGEST = "ingest"
    DRAFT = "draft"
    REVIEW = "review"
    APPROVED = "approved"
    EXPORTED = "exported"


class Product(Base):
    """A marketable product with SKU, metadata, and workflow tracking."""

    __tablename__ = "products"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    sku: Mapped[str] = mapped_column(String(100), unique=True, nullable=False)
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, default=None)
    category_id: Mapped[int | None] = mapped_column(
        Integer, ForeignKey("categories.id"), default=None
    )
    segment_id: Mapped[int | None] = mapped_column(
        Integer, ForeignKey("segments.id"), default=None
    )
    workflow_stage: Mapped[WorkflowStage] = mapped_column(
        Enum(WorkflowStage, create_constraint=True),
        default=WorkflowStage.INGEST,
        nullable=False,
    )
    is_deleted: Mapped[bool] = mapped_column(
        Boolean, default=False, nullable=False
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

    # ── Relationships ──────────────────────────────────────────────────
    category: Mapped["Category | None"] = relationship(
        "Category", back_populates="products"
    )
    segment: Mapped["Segment | None"] = relationship(
        "Segment", back_populates="products"
    )
    documents: Mapped[list["ProductDocument"]] = relationship(
        "ProductDocument", back_populates="product"
    )
    claims: Mapped[list["ProductClaim"]] = relationship(
        "ProductClaim", back_populates="product"
    )
    versions: Mapped[list["ProductVersion"]] = relationship(
        "ProductVersion", back_populates="product"
    )
    exports: Mapped[list["ExportLog"]] = relationship(
        "ExportLog", back_populates="product"
    )

File: frontend/src/types/product.ts
/**
 * Product types — PIM core entities: categories, segments, products,
 * documents, claims, versions, and related enums.
 *
 * Mirrors backend/app/models/* + backend/app/products/*_schemas.py
 */

// ── Enums ────────────────────────────────────────────────────────────────────

/** Kanban workflow stages for the content marketing pipeline. */
export const WorkflowStage = {
  INGEST: "ingest",
  DRAFT: "draft",
  REVIEW: "review",
  APPROVED: "approved",
  EXPORTED: "exported",
} as const;

export type WorkflowStage = (typeof WorkflowStage)[keyof typeof WorkflowStage];

/** Verification status for generated product claims. */
export const ClaimStatus = {
  PENDING_REVIEW: "pending_review",
  VERIFIED: "verified",
  FLAGGED: "flagged",
  REJECTED: "rejected",
} as const;

export type ClaimStatus = (typeof ClaimStatus)[keyof typeof ClaimStatus];

/** Supported document source types. */
export const DocType = {
  PDF: "pdf",
  URL: "url",
} as const;

export type DocType = (typeof DocType)[keyof typeof DocType];

// ── Category ─────────────────────────────────────────────────────────────────

export interface Category {
  id: number;
  name: string;
  description: string | null;
  product_count: number;
  created_at: string;
  updated_at: string;
}

export interface CategoryCreate {
  name: string;
  description?: string | null;
}

export interface CategoryUpdate {
  name?: string | null;
  description?: string | null;
}

// ── Segment ──────────────────────────────────────────────────────────────────

export interface Segment {
  id: number;
  name: string;
  description: string | null;
  target_audience: string | null;
  tone: string | null;
  created_at: string;
  product_count: number;
}

export interface SegmentCreate {
  name: string;
  description?: string | null;
  target_audience?: string | null;
  tone?: string | null;
}

export interface SegmentUpdate {
  name?: string | null;
  description?: string | null;
  target_audience?: string | null;
  tone?: string | null;
}

// ── ProductDocument ──────────────────────────────────────────────────────────

export interface ProductDocument {
  id: number;
  product_id: number;
  title: string;
  url: string;
  doc_type: string;
  extracted_text: string | null;
  created_at: string;
}

export interface DocumentCreate {
  url: string;
  doc_type: DocType;
}

// ── ProductClaim ─────────────────────────────────────────────────────────────

export interface ProductClaim {
  id: number;
  product_id: number;
  claim_text: string;
  source_doc_id: number | null;
  status: string;
  assigned_to: number | null;
  created_at: string;
  source_doc?: ClaimDocumentRef | null;
}

export interface ClaimDocumentRef {
  id: number;
  title: string;
  url: string;
}

export interface ClaimCreate {
  claim_text: string;
  source_doc_id?: number | null;
  status?: ClaimStatus;
}

export interface ClaimUpdate {
  claim_text?: string | null;
  status?: ClaimStatus | null;
}

// ── ProductVersion ───────────────────────────────────────────────────────────

export interface ProductVersion {
  id: number;
  version_number: number;
  snapshot_json: string;
  change_summary: string | null;
  created_by: number;
  created_at: string;
}

// ── Product ──────────────────────────────────────────────────────────────────

/** Nested category reference inside product responses. */
export interface ProductCategoryRef {
  id: number;
  name: string;
}

/** Nested segment reference inside product responses. */
export interface ProductSegmentRef {
  id: number;
  name: string;
}

/** Nested document reference inside product responses. */
export interface ProductDocumentRef {
  id: number;
  title: string;
  url: string;
  doc_type: string;
}

/** Nested claim reference inside product responses. */
export interface ProductClaimRef {
  id: number;
  claim_text: string;
  source_doc_id: number | null;
  status: string;
}

/** Flat product representation (list items). */
export interface ProductListItem {
  id: number;
  sku: string;
  name: string;
  description: string | null;
  category_id: number | null;
  segment_id: number | null;
  workflow_stage: string;
  is_deleted: boolean;
  created_at: string;
  updated_at: string;
}

/** Full product detail with nested relationships. */
export interface ProductResponse extends ProductListItem {
  category: ProductCategoryRef | null;
  segment: ProductSegmentRef | null;
  documents: ProductDocumentRef[];
  claims: ProductClaimRef[];
  versions: ProductVersion[];
}

export interface ProductCreate {
  sku: string;
  name: string;
  description?: string | null;
  category_id?: number | null;
  segment_id?: number | null;
}

export interface ProductUpdate {
  name?: string | null;
  description?: string | null;
  category_id?: number | null;
  segment_id?: number | null;
  workflow_stage?: WorkflowStage | null;
}
